Spigot-Server-Patches/0235-Anti-Xray.patch

```diff
@@ -0,0 +1,1477 @@
++    private final boolean[] solid = new boolean[Block.REGISTRY_ID.a()];
++    private final boolean[] obfuscate = new boolean[Block.REGISTRY_ID.a()];
++    //These boolean arrays represent chunk layers, true means don't obfuscate, false means obfuscate
++    private boolean[][] current = new boolean[16][16];
++    private boolean[][] next = new boolean[16][16];
++    private boolean[][] nextNext = new boolean[16][16];
++    private final DataBitsReader dataBitsReader = new DataBitsReader();
++    private final DataBitsWriter dataBitsWriter = new DataBitsWriter();
++    private final ChunkSection[] nearbyChunkSections = new ChunkSection[4];
++
++    public void obfuscate(PacketPlayOutMapChunkInfoAntiXray packetPlayOutMapChunkInfoAntiXray) {
++        boolean[] solidTemp = null;
++        boolean[] obfuscateTemp = null;
++        dataBitsReader.setDataBits(packetPlayOutMapChunkInfoAntiXray.getData());
++        dataBitsWriter.setDataBits(packetPlayOutMapChunkInfoAntiXray.getData());
++        int counter = 0;
++
++        for (int chunkSectionIndex = 0; chunkSectionIndex <= maxChunkSectionIndex; chunkSectionIndex++) {
++            if (packetPlayOutMapChunkInfoAntiXray.isWritten(chunkSectionIndex) && packetPlayOutMapChunkInfoAntiXray.getPredefinedBlockData(chunkSectionIndex) != null) {
++                int[] predefinedBlockDataBits = new int[packetPlayOutMapChunkInfoAntiXray.getPredefinedBlockData(chunkSectionIndex).length];
++
++                for (int i = 0; i < predefinedBlockDataBits.length; i++) {
++                    predefinedBlockDataBits[i] = packetPlayOutMapChunkInfoAntiXray.getDataPalette(chunkSectionIndex).a(packetPlayOutMapChunkInfoAntiXray.getPredefinedBlockData(chunkSectionIndex)[i]);
++                }
++
++                dataBitsWriter.setIndex(packetPlayOutMapChunkInfoAntiXray.getDataBitsIndex(chunkSectionIndex));
++
++                if (chunkSectionIndex == 0 || !packetPlayOutMapChunkInfoAntiXray.isWritten(chunkSectionIndex - 1)) {
++                    //Initialize some stuff if the chunk section below isn't written to the chunk packet
++                    dataBitsReader.setBitsPerValue(packetPlayOutMapChunkInfoAntiXray.getBitsPerValue(chunkSectionIndex));
++                    dataBitsReader.setIndex(packetPlayOutMapChunkInfoAntiXray.getDataBitsIndex(chunkSectionIndex));
++                    solidTemp = readDataPalette(packetPlayOutMapChunkInfoAntiXray.getDataPalette(chunkSectionIndex), solid, solidGlobal);
++                    obfuscateTemp = readDataPalette(packetPlayOutMapChunkInfoAntiXray.getDataPalette(chunkSectionIndex), obfuscate, obfuscateGlobal);
++                    //Read the blocks of the upper layer of the chunk section below if it exists
++                    ChunkSection belowChunkSection = null;
++                    boolean skipFirstLayer = chunkSectionIndex == 0 || (belowChunkSection = packetPlayOutMapChunkInfoAntiXray.getChunk().getSections()[chunkSectionIndex - 1]) == Chunk.a;
```

Comment:
since it is used a lot below too, id like to see us avoid using the obfuscated .a's for empty objects on this and the ItemStack type stuff. 

we ultimately need a public static Chunk EMPTY = a; obf helper, but that really shouldn't be in this patch. Can you add that obf helper to MC Utils patch, then utilize it here.